solr/core/src/java/org/apache/solr/filestore/FileStore.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.filestore;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+import org.apache.solr.common.MapWriter;
+import org.apache.solr.filestore.FileStoreAPI.MetaData;
+
+
+public interface FileStore {
+
+  /**Store a file into the filestore. This should ensure that it is replicated
+   * across all nodes in the cluster
+   */
+  void put(String path , MetaData metadata, ByteBuffer filecontent) throws IOException;
+
+  /** read file content from a given path
+   */
+  void get(String path, BiConsumer<InputStream, MetaData> filecontent) throws IOException;
```

Comment:
"filecontent" here refers to both content and metadata. However, for the put(), it only means the contents. 

Instead of dealing with content and metadata separately, can we instead define an abstraction class that encapsulate both the File and the metadata? E.g. 

class FileStoreFile { ByteBuffer buf, Metadata meta }

and similarly, FileStoreInputStream and FileStoreOutputStream etc.